sdk/shared/src/test/scala/org/ergoplatform/sdk/ContractTemplateSpecification.scala

```diff
@@ -0,0 +1,344 @@
+          ConstantPlaceholder(1, SType.typeByte)),
+          ConstantPlaceholder(2, SType.typeByte)).toSigmaProp
+      ),
+      t =>
+        t.isInstanceOf[IllegalArgumentException] &&
+          t.getMessage.contains("constantIndex 0 does not have a default value and absent from parameter as well"))
+  }
+
+  property("applyTemplate") {
+    val parameters = IndexedSeq(
+      createParameter("p1", 0),
+      createParameter("p2", 1),
+      createParameter("p3", 2))
+    val expressionTrees = IndexedSeq(
+      EQ(Plus(ConstantPlaceholder(0, SType.typeByte),
+        ConstantPlaceholder(1, SType.typeByte)),
+        ConstantPlaceholder(2, SType.typeByte)).toSigmaProp,
+      EQ(Plus(ConstantPlaceholder(0, SType.typeInt),
+        ConstantPlaceholder(1, SType.typeInt)),
+        ConstantPlaceholder(2, SType.typeInt)).toSigmaProp,
+      EQ(Plus(CBigInt(BigInteger.valueOf(10L)), BigIntConstant(20L)), BigIntConstant(30L)).toSigmaProp
+    )
+    val templates = Seq(
+      createContractTemplate(
+        IndexedSeq(SType.typeByte, SType.typeByte, SType.typeByte).asInstanceOf[IndexedSeq[SType]],
+        Some(IndexedSeq(Some(10.toByte), Some(20.toByte), Some(30.toByte)).asInstanceOf[IndexedSeq[Option[SType#WrappedType]]]),
+        parameters,
+        expressionTrees(0)
+      ),
+      createContractTemplate(
+        IndexedSeq(SType.typeInt, SType.typeInt, SType.typeInt).asInstanceOf[IndexedSeq[SType]],
+        Some(IndexedSeq(Some(10), None, Some(30)).asInstanceOf[IndexedSeq[Option[SType#WrappedType]]]),
+        parameters,
+        expressionTrees(1)
+      ),
+      createContractTemplate(
+        SType.EmptySeq,
+        None,
+        Parameter.EmptySeq,
+        expressionTrees(2)
+      )
+    )
+    val templateValues = Seq(
+      Map("p1" -> ByteConstant(10.toByte), "p2" -> ByteConstant(40.toByte), "p3" -> ByteConstant(50.toByte)),
+      Map("p1" -> IntConstant(10), "p2" -> IntConstant(20)),
+      Map.empty[String, Constant[SType]],
+    )
+    var expectedErgoTreeVersion = (ErgoTree.ConstantSegregationHeader | ergoTreeVersionInTests).toByte
+    if (ergoTreeVersionInTests > 0) {
+      expectedErgoTreeVersion = (expectedErgoTreeVersion | ErgoTree.SizeFlag).toByte
+    }
+    val expectedErgoTree = Seq(
+      ErgoTree(
+        expectedErgoTreeVersion,
+        IndexedSeq(
+          ByteConstant(10.toByte),
+          ByteConstant(40.toByte),
+          ByteConstant(50.toByte)
+        ),
+        expressionTrees(0)
+      ),
+      ErgoTree(
+        expectedErgoTreeVersion,
+        IndexedSeq(
+          IntConstant(10),
+          IntConstant(20),
+          IntConstant(30)
+        ),
+        expressionTrees(1)
+      ),
+      ErgoTree(
+        expectedErgoTreeVersion,
+        Constant.EmptySeq,
+        expressionTrees(2)
+      )
+    )
+
+    templates.indices.foreach(i =>
+      templates(i).applyTemplate(Some(ergoTreeVersionInTests), templateValues(i)) shouldEqual expectedErgoTree(i)
+    )
+  }
+
```

Comment:
One problem with the tests below. They are not systematic. There two ways to solve it. 1) create template generator 2) manually create many templates from simples to more complex increasing complexity in one parameter on each step.
If do it manually, it would be easier to reuse common sub-components. It is safe becase the data structure is ummutable so parts can be shared by many instances.

Reply:
Done, went ahead with the contract template generator approach.